gaeguli/pipeline.c

```diff
@@ -377,6 +379,8 @@ _link_probe_cb (GstPad * pad, GstPadProbeInfo * info, gpointer user_data)
   g_autoptr (GstPad) sink_pad = NULL;
   g_autoptr (GaeguliPipeline) self = g_object_ref (link_target->self);
 
+  gst_pad_remove_probe (pad, info->id);
```

Comment:
My problem with GST_PAD_PROBE_TYPE_IDLE was that the callback may get invoked immediately from inside gst_pad_add_probe() in the main thread.

In the tests, pipeline and fifo transmit share the same thread, so running the probe function to remove a pipeline inside that main thread will pause reading from fifos. If at that moment the removed pipeline's filesink is blocked inside a write() call, waiting for free space in the fifo to become available, the pipeline's state transition from PLAYING to NULL will take forever, causing a deadlock.

Changing probe type to BLOCK moves _link_probe_cb() invocation into some GStreamer's internal thread. This frees the main thread, which then keeps reading from the fifo, so that filling up the buffer can't block the fifo's writing end indefinitely.

A downside of BLOCK probe is that if no data or events are flowing through the pad, the callback won't get invoked at all. This doesn't seem to be our case as we always have some live buffers coming from the camera. Should it become a problem, we can always send EOS event to the pad to have the probe function called in absence of data.

> Plus, this callback will be removed by returning GST_PAD_PROBE_REMOVE. Is it fully safe from doubly-removed?

We must remove the probe before we call gst_ghost_pad_new(); that function sends new events on the blocked pad, which would invoke our probe again and we don't want that. Looking at GstPad's probe_hook_marshal(), removing the probe from inside its callback is safe and GST_PAD_PROBE_REMOVE has no effect then.

Reply:
That's very valuable information. Thank you.